File: src/python/MetaphorLexer.py
from Token import Token, TokenType
from Lexer import Lexer

class MetaphorLexer(Lexer):
    """
    Lexer for handling the metaphor language with its specific syntax, including keywords like
    Action, Context, Role, and proper indentation handling.
    """
    def __init__(self, filename, indent_spaces=4):
        """
        Initializes the MetaphorLexer.

        Args:
            filename (str): The filename to be lexed.
            indent_spaces (int): The number of spaces that make up one level of indentation
                (default is 4).
        """
        self.keyword_map = {
            "Include:": TokenType.INCLUDE,
            "Embed:": TokenType.EMBED,
            "Action:": TokenType.ACTION,
            "Context:": TokenType.CONTEXT,
            "Role:": TokenType.ROLE
        }
        self.in_text_block = False
        self.indent_column = 1
        self.indent_spaces = indent_spaces
        super().__init__(filename)

    def _tokenize(self):
        """Tokenizes the input file into appropriate tokens."""
        lines = self.input.splitlines()
        for line in lines:
            self._process_line_contents(line)
            self.current_line += 1

        # Handles remaining outdents as the file has ended.
        while self.indent_column > 1:
            self.tokens.append(Token(TokenType.OUTDENT, "[Outdent]", "", self.filename,
                                        self.current_line, self.indent_column))
            self.indent_column -= self.indent_spaces

    def _process_indentation(self, line, start_column):
        """Processes the indentation of the current line."""
        if len(line) == 0:
            return

        # Calculate the difference in indentation
        indent_offset = start_column - self.indent_column

        # Handle indentation increase (INDENT)
        if indent_offset > 0:
            if indent_offset % self.indent_spaces != 0:
                self.tokens.append(Token(TokenType.BAD_INDENT, "[Bad Indent]", line,
                                            self.filename, self.current_line,
                                            start_column))
                return

            while indent_offset > 0:
                self.tokens.append(Token(TokenType.INDENT, "[Indent]", line, self.filename,
                                            self.current_line, start_column))
                indent_offset -= self.indent_spaces

            self.indent_column = start_column

        # Handle indentation decrease (OUTDENT)
        elif indent_offset < 0:
            if abs(indent_offset) % self.indent_spaces != 0:
                self.tokens.append(Token(TokenType.BAD_OUTDENT, "[Bad Outdent]", line,
                                            self.filename, self.current_line, start_column))
                return

            while indent_offset < 0:
                self.tokens.append(Token(TokenType.OUTDENT, "[Outdent]", line, self.filename,
                                            self.current_line, start_column))
                indent_offset += self.indent_spaces

            self.indent_column = start_column

    def _process_line_contents(self, line):
        """
        Read the next line and determine if it's a keyword followed by text.
        It also processes indentation at the start of the line.

        Args:
            line (str): The current line to process.

        Returns:
            list[Token]: A list of Token instances representing both indentation and text.
        """
        stripped_line = line.lstrip(' ')
        start_column = len(line) - len(stripped_line) + 1

        if stripped_line:
            # Is this a comment?  If yes, we're done.
            if stripped_line.startswith("#"):
                return

            # Split the line by spaces to check for a keyword followed by text
            words = stripped_line.split(maxsplit=1)

            # Check if the first word is a recognized keyword
            if words[0] in self.keyword_map:
                # Create a keyword token
                self._process_indentation(line, start_column)
                self.tokens.append(Token(self.keyword_map[words[0]], words[0], line,
                                            self.filename, self.current_line, start_column))

                # If there is text after the keyword, create a separate text token
                if len(words) > 1:
                    self.tokens.append(Token(TokenType.KEYWORD_TEXT, words[1], line,
                                                self.filename, self.current_line,
                                                start_column + len(words[0]) + 1))

                self.in_text_block = False
                return

        # We're dealing with text.  If we're already in a text block then we want to use the same
        # indentation level for all rows of text unless we see outdenting (in which case we've got
        # bad text, but we'll leave that to the parser).
        if self.in_text_block:
            if start_column > self.indent_column:
                start_column = self.indent_column
            elif start_column < self.indent_column:
                self._process_indentation(line, start_column)
        else:
            self._process_indentation(line, start_column)

        # If no keyword is found, treat the whole line as text
        text_line = line[start_column - 1:]
        if self.in_text_block or len(text_line) > 0:
            self.tokens.append(Token(TokenType.TEXT, line[start_column - 1:], line,
                                        self.filename, self.current_line, start_column))

        if len(text_line) > 0:
            self.in_text_block = True

File: src/m6rc/MetaphorLexer.cpp
#include <cctype>
#include <iostream>
#include <fstream>

#include "MetaphorLexer.hpp"

#define INDENT_SPACES 4

MetaphorLexer::MetaphorLexer(const std::string& filename) :
        Lexer(filename),
        indentColumn_(1),
        processingIndent_(false),
        inTextBlock_(false) {
    lexTokens();
}

auto MetaphorLexer::processIndentation(size_t column) -> void {
    int indentOffset = column - indentColumn_;
    if (!indentOffset) {
        return;
    }

    if ((indentOffset % INDENT_SPACES) != 0) {
        if (indentOffset > 0) {
            tokens_.push_back(Token(TokenType::BAD_INDENT, "[Bad indent]", line_, filename_, currentLine_, column));
            return;
        }

        tokens_.push_back(Token(TokenType::BAD_OUTDENT, "[Bad outdent]", line_, filename_, currentLine_, column));
        return;
    }

    indentColumn_ += indentOffset;
    if (indentOffset >= INDENT_SPACES) {
        while (indentOffset) {
            indentOffset -= INDENT_SPACES;
            tokens_.push_back(Token(TokenType::INDENT, "[Indent]", line_, filename_, currentLine_, column));
        }

        return;
    }

    while (indentOffset) {
        indentOffset += INDENT_SPACES;
        tokens_.push_back(Token(TokenType::OUTDENT, "[Outdent]", line_, filename_, currentLine_, column));
    }
}

auto MetaphorLexer::consumeWhitespace() -> void {
    while (position_ < input_.size() && isspace(input_[position_]) && input_[position_] != '\n') {
        position_++;
        currentColumn_++;
    }
}

auto MetaphorLexer::readKeywordOrText() -> void {
    int startColumn = currentColumn_;
    if (processingIndent_) {
        processIndentation(startColumn);
        processingIndent_ = false;
    }

    size_t startPosition = position_;
    while (position_ < input_.size() && !isspace(input_[position_]) && input_[position_] != '\n') {
        position_++;
        currentColumn_++;
    }

    std::string word = input_.substr(startPosition, position_ - startPosition);

    // If we have a keyword then return that.
    if (keyword_map.find(word) != keyword_map.end()) {
        // Once we've seen a keyword, we're no longer in a text block.
        inTextBlock_ = false;
        tokens_.push_back(Token(keyword_map.at(word), word, line_, filename_, currentLine_, startColumn));
        return;
    }

    // Have we already seen a keyword?  If yes then this is keyword text
    if (seenNonWhitespaceCharacters_) {
        position_ = endOfLine_;
        tokens_.push_back(Token(TokenType::KEYWORD_TEXT, line_.substr(startColumn - 1, endOfLine_ - startOfLine_ - (startColumn - 1)), line_, filename_, currentLine_, startColumn));
        return;
    }

    // We're dealing with text.  If we're already in a text block then we want to use the same indentation
    // level for all rows of text unless we see outdenting (in which case we've got bad text, but we'll
    // leave that to the parser).
    if (inTextBlock_) {
        if (startColumn > indentColumn_) {
            startColumn = indentColumn_;
        }
    }

    inTextBlock_ = true;
    position_ = endOfLine_;
    tokens_.push_back(Token(TokenType::TEXT, line_.substr(startColumn - 1, endOfLine_ - startOfLine_ - (startColumn - 1)), line_, filename_, currentLine_, startColumn));
}

auto MetaphorLexer::lexTokens() -> void {
    // Get the next token.
    while (position_ < input_.size()) {
        char ch = input_[position_];

        // If we have a new line then get the next one.
        if (ch == '\n') {
            // If we've not seen any non-whitespace characters and we're in a text block then emit a blank
            // line.  Then pretend we saw characters so next time we process the end of line.
            if (!seenNonWhitespaceCharacters_ && inTextBlock_) {
                tokens_.push_back(Token(TokenType::TEXT, "", line_, filename_, currentLine_, indentColumn_));
            }

            processingIndent_ = true;
            consumeNewline();
            updateEndOfLine();
            seenNonWhitespaceCharacters_ = false;
            continue;
        }

        if (isspace(ch)) {
            consumeWhitespace();
            continue;
        }

        // If we have a comment then skip over everything until the end of the current line.
        if (ch == '#') {
            position_ = endOfLine_;
            continue;
        }

        readKeywordOrText();
        seenNonWhitespaceCharacters_ = true;
    }

    tokens_.push_back(Token(TokenType::END_OF_FILE, "", line_, filename_, currentLine_, 1));
}
